### PR TITLE
mgr/cephadm: redeploy nfs daemons in error state w/ backoff

### PR DESCRIPTION
**Pull request by adk3798**
In order to keep up the availability of nfs daemons we want to try to keep them in a running state. Right now, we are  already able to fix nfs daemons that get removed somehow because the daemon will get put back when applying service specs. However, an nfs daemon that still exists but is in an error state is left in that state. The idea of this change is to try to "fix" these daemons by redeploying them.

The backoff portion is to try to stop us from repeatedly trying to redeploy a daemon that for whatever reason cannot be repaired. After a repair attempt, we wait a certain amount of time before we are allowed another attempt for that daemon. This amount of time increases with each subsequent attempt until we hit a point where we just give up.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/inventory.py [hunk 1]
```diff
@@ -1082,6 +1084,38 @@ def get_scheduled_daemon_action(self, host: str, daemon: str) -> Optional[str]:
 
         return self.scheduled_daemon_actions.get(host, {}).get(daemon)
 
+    def can_repair(self, daemon_name: str) -> bool:
```
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py]: using "redeploy_on_error" in these names now

File: src/pybind/mgr/cephadm/inventory.py [hunk 2]
```diff
@@ -1082,6 +1084,38 @@ def get_scheduled_daemon_action(self, host: str, daemon: str) -> Optional[str]:
...
+        backoff = {
+            1: 90,  # 90 seconds
+            2: 180,  # 3 minutes
+            3: 300,  # 5 minutes
+            4: 600,  # 10 minutes
+            5: 1800,  # 30 minutes
+            6: 7200,  # 2 hours
+            7: 43200,  # 12 hours
+            8: 172800,  # 2 days
+        }
```
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py]: @mgfritch is what I put in now sort of what you had in mind or was it something different?

File: src/pybind/mgr/cephadm/utils.py [hunk 3]
```diff
@@ -133,3 +136,10 @@ def file_mode_to_str(mode: int) -> str:
             f'{"x" if (mode >> shift) & 1 else "-"}'
         ) + r
     return r
+
+
+def exponential_backoff(t: int) -> int:
+    if t < 1:  # default non-positive ints to result for t=1
+        return 90
+    # 90s, 6m, 24m, 1h|36m, 6h|24m, 1d|1h|36m, etc.
+    return (4 ** t) * 90
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/utils.py:145]:
```suggestion
def exponential_backoff(try_num: int = 0, sleep_interval: int = 90) -> int:
    if try_num < 0:
        return 90
    # 90s, 3m, 6m ..
    return (2 ** tries) * sleep_interval
```
make the interval more frequent and add these as params?

**Review by mgfritch — COMMENTED** [review]
should we eventually give up on the redeploy after a certain number of retries?

**Comment by ajarr** [comment]
> This PR is imo scary.
> 
> 1. we're re-implementing systemd's automatic restart of units. Is there a way we can just trust the systemd implementation?

Good point! I don't understand why ganesha server daemon needs a special redeploy. What happens when any other ceph server daemon managed by cephadm gets into an error state? I couldn't get this info from the commit message.

> 2. we have to be extraordinary careful that this backoff doesn't interfere with other things, like the upgrade
> 3. users need a way to manually clear the backoff.

**Comment by adk3798** [comment]
Rethought all this, cases where this would fix the daemon but systemd redeploy wouldn't are few and far between. Not worth the added complexity.

**Pull request closed without merging**